dynamic_array.c

```diff
@@ -0,0 +1,340 @@
+#include "dynamic_array.h"
+
+// NOTE: When we destroy items, we call the destructor for them 
+// iff the destructor is not NULL.
+
+typedef void (*destr)(void *);
+
+// Initialiser - sets fields in instance and allocates buffer (via buffer_alloc)
+void buffer_init(struct buffer *inst, size_t item_size, size_t capacity, 
+                 size_t alloc_by, void (*destructor)(void *))
+{   
+    inst->capacity   = 0;
+    inst->length     = 0;
+    inst->item_size  = item_size;
+    inst->alloc_by   = alloc_by;
+    inst->data       = NULL;
+    inst->destructor = destructor;
+    inst->buffer     = NULL;
+
+    buffer_alloc(inst, capacity);
+}
+
+// Destructor - frees the buffer, calling the destructor for each item in the buffer
+void buffer_destroy(struct buffer *inst)
+{
+    for(int i = 0; i < inst->capacity; i++){
+        ((destr)(inst->destructor))(inst->buffer[i]);
```

Comment:
Casting functions when calling them is generally a symptom of design issues - I've already commented earlier on the typing for destructors.  When using the new "destructor *" type for destructor function pointers, this cast will not be necessary.